File: roms/strategy_x.py
#
#	Strategy X
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    prg1 = z.read('2c_1.bin') + z.read('2e_2.bin') + z.read('2f_3.bin') + z.read('2h_4.bin') + z.read('2j_5.bin') + z.read('2l_6.bin')
    prg2 = z.read('s1.bin') + z.read('s2.bin')
    bg = z.read('5f_c2.bin') + z.read('5h_c1.bin')
    rgb = z.read('strategy.6e')
    map = z.read('strategy.10k')

rom = prg1 + prg2 + bg + rgb + map

def pngstring(a):
    w = 1024
    img = Image.new('P', (w, ceil(len(a) / w)))
    img.putpalette(sum([[i, 0, 0] for i in range(256)], []))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[2], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/strategy_x.js
/*
 *	Strategy X
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};

const z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg1 = cat(z.read('2c_1.bin'), z.read('2e_2.bin'), z.read('2f_3.bin'), z.read('2h_4.bin'), z.read('2j_5.bin'), z.read('2l_6.bin'));
const prg2 = cat(z.read('s1.bin'), z.read('s2.bin'));
const bg = cat(z.read('5f_c2.bin'), z.read('5h_c1.bin'));
const rgb = z.read('strategy.6e');
const map = z.read('strategy.10k');

const rom = cat(prg1, prg2, bg, rgb, map);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[3], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
